File: backend/src/api/icd_search.rs
use actix_web::{web, HttpResponse, Result};
use serde::{Deserialize, Serialize};
use crate::codecs::icd::{IcdCodec, IcdFilter, IcdDiscipline};

// Basic response structure
#[derive(Serialize, Deserialize)]
pub struct ApiResponse {
    pub service: String,
    pub status: String,
    pub message: String,
    pub timestamp: String,
}

// ICD search endpoint
pub async fn icd_search(
    query: web::Query<std::collections::HashMap<String, String>>
) -> Result<HttpResponse> {
    let codec = IcdCodec::new();
    let discipline = query.get("discipline")
        .and_then(|d| match d.to_lowercase().as_str() {
            "biomedicine" => Some(IcdDiscipline::Biomedicine),
            "tm2" => Some(IcdDiscipline::TM2),
            _ => None,
        });

    let filter = IcdFilter {
        discipline,
        search_term: query.get("search").cloned(),
        parent_filter: query.get("parent").cloned(),
    };

    match codec.search_codes(filter, query.get("limit").and_then(|l| l.parse().ok())).await {
        Ok(codes) => {
            let formatted = codec.format_response(codes);
            Ok(HttpResponse::Ok().json(serde_json::json!({
                "service": "ICD-11 Search",
                "total": formatted.len(),
                "results": formatted,
                "timestamp": chrono::Utc::now().to_rfc3339()
            })))
        },
        Err(e) => Ok(HttpResponse::InternalServerError().json(ApiResponse {
            service: "ICD-11 Search".to_string(),
            status: "error".to_string(),
            message: format!("Search failed: {}", e),
            timestamp: chrono::Utc::now().to_rfc3339(),
        }))
    }
}

// ICD all codes endpoint
pub async fn icd_all(
    query: web::Query<std::collections::HashMap<String, String>>
) -> Result<HttpResponse> {
    let codec = IcdCodec::new();
    match codec.get_all_codes(query.get("limit").and_then(|l| l.parse().ok())).await {
        Ok(codes) => {
            let formatted = codec.format_response(codes);
            Ok(HttpResponse::Ok().json(serde_json::json!({
                "service": "ICD-11 All Codes",
                "total": formatted.len(),
                "results": formatted,
                "timestamp": chrono::Utc::now().to_rfc3339()
            })))
        },
        Err(e) => Ok(HttpResponse::InternalServerError().json(ApiResponse {
            service: "ICD-11 All Codes".to_string(),
            status: "error".to_string(),
            message: format!("Failed: {}", e),
            timestamp: chrono::Utc::now().to_rfc3339(),
        }))
    }
}

// Get biomedicine codes
pub async fn icd_biomedicine(
    query: web::Query<std::collections::HashMap<String, String>>
) -> Result<HttpResponse> {
    let codec = IcdCodec::new();
    match codec.get_biomedicine_codes(query.get("limit").and_then(|l| l.parse().ok())).await {
        Ok(codes) => {
            let formatted = codec.format_response(codes);
            Ok(HttpResponse::Ok().json(serde_json::json!({
                "service": "ICD-11 Biomedicine",
                "discipline": "BIOMEDICINE",
                "total": formatted.len(),
                "results": formatted,
                "timestamp": chrono::Utc::now().to_rfc3339()
            })))
        },
        Err(e) => Ok(HttpResponse::InternalServerError().json(ApiResponse {
            service: "ICD-11 Biomedicine".to_string(),
            status: "error".to_string(),
            message: format!("Failed: {}", e),
            timestamp: chrono::Utc::now().to_rfc3339(),
        }))
    }
}

// Get TM2 codes
pub async fn icd_tm2(
    query: web::Query<std::collections::HashMap<String, String>>
) -> Result<HttpResponse> {
    let codec = IcdCodec::new();
    match codec.get_tm2_codes(query.get("limit").and_then(|l| l.parse().ok())).await {
        Ok(codes) => {
            let formatted = codec.format_response(codes);
            Ok(HttpResponse::Ok().json(serde_json::json!({
                "service": "ICD-11 Traditional Medicine",
                "discipline": "TM2",
                "total": formatted.len(),
                "results": formatted,
                "timestamp": chrono::Utc::now().to_rfc3339()
            })))
        },
        Err(e) => Ok(HttpResponse::InternalServerError().json(ApiResponse {
            service: "ICD-11 TM2".to_string(),
            status: "error".to_string(),
            message: format!("Failed: {}", e),
            timestamp: chrono::Utc::now().to_rfc3339(),
        }))
    }
}

File: backend/app/api/icd_search.py
from datetime import datetime, timezone
from typing import Optional

from fastapi import APIRouter
from fastapi.responses import JSONResponse

from app.codecs.icd import IcdCodec, IcdDiscipline, IcdFilter

router = APIRouter()


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def _parse_discipline(value: Optional[str]) -> Optional[IcdDiscipline]:
    if value is None:
        return None
    lowered = value.lower()
    if lowered == "biomedicine":
        return IcdDiscipline.BIOMEDICINE
    if lowered == "tm2":
        return IcdDiscipline.TM2
    return None


@router.get("/search")
async def icd_search(
    search: Optional[str] = None,
    discipline: Optional[str] = None,
    parent: Optional[str] = None,
    limit: Optional[int] = None,
) -> JSONResponse:
    codec = IcdCodec()
    filter = IcdFilter(
        discipline=_parse_discipline(discipline),
        search_term=search,
        parent_filter=parent,
    )

    try:
        codes = await codec.search_codes(filter, limit)
        formatted = codec.format_response(codes)
        return JSONResponse(
            {
                "service": "ICD-11 Search",
                "total": len(formatted),
                "results": formatted,
                "timestamp": _now(),
            }
        )
    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={
                "service": "ICD-11 Search",
                "status": "error",
                "message": f"Search failed: {e}",
                "timestamp": _now(),
            },
        )


@router.get("/all")
async def icd_all(limit: Optional[int] = None) -> JSONResponse:
    codec = IcdCodec()
    try:
        codes = await codec.get_all_codes(limit)
        formatted = codec.format_response(codes)
        return JSONResponse(
            {
                "service": "ICD-11 All Codes",
                "total": len(formatted),
                "results": formatted,
                "timestamp": _now(),
            }
        )
    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={
                "service": "ICD-11 All Codes",
                "status": "error",
                "message": f"Failed: {e}",
                "timestamp": _now(),
            },
        )


@router.get("/biomedicine")
async def icd_biomedicine(limit: Optional[int] = None) -> JSONResponse:
    codec = IcdCodec()
    try:
        codes = await codec.get_biomedicine_codes(limit)
        formatted = codec.format_response(codes)
        return JSONResponse(
            {
                "service": "ICD-11 Biomedicine",
                "discipline": "BIOMEDICINE",
                "total": len(formatted),
                "results": formatted,
                "timestamp": _now(),
            }
        )
    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={
                "service": "ICD-11 Biomedicine",
                "status": "error",
                "message": f"Failed: {e}",
                "timestamp": _now(),
            },
        )


@router.get("/tm2")
async def icd_tm2(limit: Optional[int] = None) -> JSONResponse:
    codec = IcdCodec()
    try:
        codes = await codec.get_tm2_codes(limit)
        formatted = codec.format_response(codes)
        return JSONResponse(
            {
                "service": "ICD-11 Traditional Medicine",
                "discipline": "TM2",
                "total": len(formatted),
                "results": formatted,
                "timestamp": _now(),
            }
        )
    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={
                "service": "ICD-11 TM2",
                "status": "error",
                "message": f"Failed: {e}",
                "timestamp": _now(),
            },
        )
